### PR TITLE
CRM-16526 [needs-review] Financial ACL API tests 

### PR DESCRIPTION
**Pull request by Edzelopez**
- [CRM-16526: ACLs for Financial Types](https://issues.civicrm.org/jira/browse/CRM-16526)

### CODE REVIEW

File: api/v3/LineItem.php [hunk 1]
```diff
@@ -44,8 +44,28 @@
  *   api result array
  */
 function civicrm_api3_line_item_create($params) {
-  $params = CRM_Contribute_BAO_Contribution::checkTaxAmount($params, TRUE);
-  return _civicrm_api3_basic_create(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus()) {
```
Comment (eileenmcnaughton) [inline api/v3/LineItem.php]: Can't this go in the BAO? the api should just be a wrapper for the BAO

**Comment by eileenmcnaughton** [comment]
Great to see the test!

What is worrying me about all this extra logic going into the api layer is that apart from being against our general principle of trying to put all logic in the BAO layer you are also creating future-you or someone like you work because api v4 will drop all of the api layer create logic and you will either have to move it to the BAO at that point or negotiate some exception on api v4.

Of course moving it to the BAO is unit tested but it would require at least some UI click around checks too to check nothing wierd is happening.

**Comment by Edzelopez** [comment]
I agree @eileenmcnaughton, but I checked and the lineitem get API does not rely on the BAO. It just does a basic DAO get (). So unless we change that, we can't really modify the way the line items are being retrieved. Of course if there's anot her way this can be done please let me know :)

**Comment by eileenmcnaughton** [comment]
I was looking more at create than get (which does have a BAO function to add things to).

The issue with get leveraging ACLs has recently been looked at by @colemanw and he can hopefully advise as to whether that approach is mature enough to add here.

I'm also kind of aware that you used an ACL approach that is more drupal-like than CiviCRM-like but I didn't dig into that side of it - I suspect the difference won't have much impact thought

**Comment by Edzelopez** [comment]
Thanks for the suggestion @eileenmcnaughton 
I've added the check to perform at BAO level for create rather than API, although I have set financial_type_id to be a required field for creating/editing a lineitem.

File: CRM/Price/BAO/LineItem.php [hunk 2]
```diff
@@ -67,6 +69,12 @@ public static function create(&$params) {
     if ($id) {
       unset($params['entity_id'], $params['entity_table']);
     }
+    if (CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus()) {
+      CRM_Financial_BAO_FinancialType::getAvailableFinancialTypes($types, $op);
+      if (!in_array($params['financial_type_id'], array_keys($types))) {
```
Comment (eileenmcnaughton) [inline CRM/Price/BAO/LineItem.php]: You need to only do this is check_permissions = TRUE

File: api/v3/LineItem.php [hunk 3]
```diff
@@ -75,6 +76,23 @@ function _civicrm_api3_line_item_create_spec(&$params) {
  *   Array of matching line_items
  */
 function civicrm_api3_line_item_get($params) {
+  if (CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus()) {
```
Comment (eileenmcnaughton) [inline api/v3/LineItem.php]: Make sure you only do checks when check_permission is set

File: api/v3/LineItem.php [hunk 4]
```diff
@@ -90,5 +108,17 @@ function civicrm_api3_line_item_get($params) {
...
+    CRM_Financial_BAO_FinancialType::getAvailableFinancialTypes($types, CRM_Core_Action::DELETE);
+    if (empty($params['financial_type_id'])) {
+      $params['financial_type_id'] = civicrm_api3('LineItem', 'getvalue', array(
+        'id' => $params['id'],
+        'return' => 'financial_type_id',
+      ));
+    }
+    if (!in_array($params['financial_type_id'], array_keys($types))) {
+      throw new API_Exception('You do not have permission to delete this line item');
```
Comment (eileenmcnaughton) [inline api/v3/LineItem.php]: Ditto

**Comment by Edzelopez** [comment]
Jenkins, retest this please

**Comment by Edzelopez** [comment]
Jenkins, retest this please

**Comment by pradpnayak** [comment]
Hi Eileen,

We are having an issue with getting Jenkins to pass.
When running all tests for api_v3_FinancialTypeACLTest via PHPUnit on our local machines, we get a successful assertion after the tests have finished completion. Unfortunately, this is not the same when we try to merge the PR to master and let Jenkins test the API test there. Such can be seen from here.

We were wondering if you have any idea what we are doing wrong here and would very much appreciate ideas as to how to make sure that there is consistency of our tests between our local machine and Jenkins.

**Comment by eileenmcnaughton** [comment]
See if it works when running api_v3_AllTests - probably some other test is leaving something behind that is messing with your test so it fails when they all run but not with only some. If it does fail there is a clever way to run selectively but I just delete around half the test classes locally & re-run until I've homing in on the clashing class

**Comment by Edzelopez** [comment]
@eileenmcnaughton you were right, there was a caching issue with one of the static variables that was failing when run with the other tests. I've fixed it now.

**Comment by JoeMurray** [comment]
@eileenmcnaughton would you be able to merge this for us? It's been waiting for over a week.

**Pull request merged**